addons/web/static/src/core/virtual_hook.js

```diff
@@ -0,0 +1,119 @@
+ * @property {PixelValue | (item: T) => PixelValue} [itemHeight=0]
+ * @property {PixelValue} [margin="100%"]
+ */
+
+/**
+ * @typedef ScrollPosition
+ * @property {number} top
+ */
+
+/** @typedef {number | `${string}px` | `${string}%`} PixelValue */
+
+/**
+ * Converts a number,
+ *
+ * @param {PixelValue} value
+ * @returns {number}
+ */
+const toPixels = (value) => {
+    if (typeof value === "number") {
+        return value;
+    }
+    if (value.endsWith("%")) {
+        return window.innerHeight * (Number(value.slice(0, -1)) / 100);
+    }
+    if (value.endsWith("px")) {
+        return Number(value.slice(0, -2));
+    }
+};
+
+/**
+ * Calculates the displayed items in a virtual list.
+ *
+ * @template T
+ * @param {VirtualHookParams<T>} params
+ * @returns {ReturnType<useState<T>>}
+ */
+export function useVirtual({ items, scrollableRef, initialScroll, itemHeight, margin }) {
+    const computeVirtualItems = () => {
+        const { items, scroll } = current;
+
+        const yMargin = toPixels(marginValue);
+        const yStart = scroll.top - yMargin;
+        const yEnd = scroll.top + window.innerHeight + yMargin;
+
+        let [startIndex, endIndex] = [0, 0];
+        let currentTop = 0;
+
+        for (const item of items) {
+            const height = toPixels(getItemHeight(item));
+            if (currentTop + height < yStart) {
+                startIndex++;
+                endIndex++;
+            } else if (currentTop + height <= yEnd + height) {
+                endIndex++;
+            } else {
+                break;
+            }
+            currentTop += height;
+        }
+
+        const prevItems = toRaw(virtualItems);
+        const newItems = items.slice(startIndex, endIndex);
+
+        if (!shallowEqual(prevItems, newItems)) {
+            virtualItems.length = 0;
+            virtualItems.push(...newItems);
+        }
+    };
+
+    const getItems = typeof items === "function" ? items : () => items;
+
+    const getItemHeight =
+        typeof itemHeight === "function" ? itemHeight : () => itemHeight || "100%";
+
+    const marginValue = margin === undefined ? "100%" : margin;
+
+    const current = {
+        items: getItems(),
+        scroll: { top: 0, ...initialScroll },
+    };
+
+    const virtualItems = useState([]);
+
+    onWillStart(computeVirtualItems);
+    onWillRender(() => {
+        const previousItems = current.items;
+        current.items = getItems();
+        if (!shallowEqual(previousItems, current.items)) {
+            computeVirtualItems();
+        }
+    });
+    const throttledOnScroll = throttleForAnimation((/** @type {Event} */ ev) => {
+        current.scroll.top = ev.target.scrollTop;
+        computeVirtualItems();
+    });
+    useEffect(
+        (el) => {
+            if (el) {
+                el.addEventListener("scroll", throttledOnScroll);
+                return () => el.removeEventListener("scroll", throttledOnScroll);
```

Comment:
is scroll triggered on zoom in/out?

Reply:
good remark, we'll check ;-)

Reply:
It was indeed not working:
fixed and a test has been added "updates on resize", thanks